keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,951 @@
+
+Follow-up goals:
+* When a feature is about a new field, remove it from OpenAPI specs when the
+  feature is off.
+* For features that possibly could support it, add a canarying mechanism so that
+  the feature enablement doesn't break the whole cluster at once when there is a
+  problem.
+
+### Non-Goals
+
+* The feature system will not permit passing configuration, since that's already
+  done in a variety of ways today. Therefore only features that don't take
+  config can be dynamic. Maybe this can be addressed in a follow up.
+
+## Proposal
+
+### High-level design
+
+First we will discuss all the pieces with minimal detail. Probably this section
+will leave you with questions which hopefully are answered in the "detailed
+design" section.
+
+#### Client Flow
+
+Let's call the current use of feature gates the "client flow". Things that are
+already handled correctly by the existing client flow we will try not to change.
+
+The main thing we want to add to the client flow is reporting of the feature's
+uses: did the client check the feature state, and what value did it find? The
+necessity of this will be clear when we discuss the "server flow" later.
+
+To make our improvements to the client flow, we will add a new implementation of
+the existing k8s.io/component-base/featuregate package's MutableFeatureGate
+type. It will require post-construction initialization, so that it can write the
+uses to the api; to be a drop-in replacement, it can't require an api client
+when it is constructed, because that happens before the command line is parsed /
+config files are read.
+
+Our new client will use the existing MutableFeatureGate to read the command
+line, so perhaps it should be thought of as a layer.
+
+When we update this KEP to solve the secondary goals, we will add here a
+mechanism for reading feature on/off values from the API -- that's not easy, due
+to the startup order.
+
+#### API
+
+Obviously, we will add an API object, Feature (details below). Features will be
+namespaced; this will permit all Kubernetes features to be in the kube-system
+namespace, and third parties can use another namespace.
+
+The Feature object will record, per-feature, its configured desired value,
+default value, uses, and current version. Most importantly, it will record the
+current state of the feature. There will be at least four states: On, Off,
+TurningOn, and TurningOff.
+
+Various tasks might need to be performed before exiting a transitional state;
+these will also be tracked in the API object.
+
+It is the responsibility of the server flow to initiate and complete state
+transitions.
+
+#### Server flow
+
+To explain what the server flow does, we will first assume for this overview
+that all servers are at the same version. In Kubernetes, the server flow logic
+will run in kube-apiserver. We will go through its responsibilities one at a
+time.
+
+Selecting a leader. Since multiple apiservers are running, we will pick one to
```

Comment:
👀

Reply:
If there are multi apiservers among which one is misconfigured(with a different configuration), will it be a problem?

Reply:
Added story describing what happens.

Reply:
During cluster upgrade, there may be one v1.n apiserver and two v1.(n-1) apiservers, which is preferred to be the leader?

Reply:
Great observation -- that situation needs to trigger version change functions, it is complicated and I haven't written that flow into the KEP yet. The leader in this case is per-version, since the former version needs to run pre-upgrade hooks and the destination version needs to run post-upgrade hooks. I don't have much more detail than that yet, I was focusing on getting the enablement/disablement flow right first, since that's complicated enough.